File: Array/medium/cpp/majority_element_nby2_times.cpp
//
// Created by santosh on 12/20/25.
//

#include<bits/stdc++.h>
using namespace std;

class Solution
{
    public:
    int majorityElement(vector<int>& nums)
    {
        int count = 0;
        int majorityElement=nums[0];

        for (int i=0; i<nums.size(); i++)
        {
            if (count==0)
            {
                majorityElement=nums[i];
            }
            if (nums[i]==majorityElement)
            {
                count++;
            } else
            {
                count--;
            }
        }

        for (int i=0; i<nums.size(); i++)
        {
            if (nums[i]==majorityElement)
            {
                count++;
            }
        }

        if (count>nums.size()/2)
        {
            return majorityElement;
        }

        return -1;
    }
};

int main ()
{
    // vector<int> nums={7, 0, 0, 1, 7, 7, 2, 7, 7};
    vector<int> nums={1, 1, 1, 2, 1, 2};
    Solution s;

    cout<<s.majorityElement(nums)<<endl;

}

File: Array/medium/python/majority_element_nby2_times.py
# Majority element appearing more than n/2 times (LeetCode 169)
# Boyer-Moore Voting Algorithm
# Created by santosh

class Solution:
    def majority_element(self, nums: list[int]) -> int:
        """
        Boyer-Moore Voting Algorithm.
        Time: O(n), Space: O(1)
        """
        count = 0
        candidate = nums[0]

        for num in nums:
            if count == 0:
                candidate = num
            count += 1 if num == candidate else -1

        # Verify the candidate
        count = sum(1 for num in nums if num == candidate)
        if count > len(nums) // 2:
            return candidate

        return -1


if __name__ == "__main__":
    nums = [1, 1, 1, 2, 1, 2]
    s = Solution()
    print(s.majority_element(nums))
